homework-5/homework-5.py

```diff
@@ -0,0 +1,86 @@
+# Backend Programming with Python
+# Homework 5 - Battleships
+
+import random, math
+
+
+class SpaceShip:
+    def __init__(self, name, hp, armour, damage):
+        self.name = name
+        self.hp = hp
+        self.armour = armour
+        self.damage = damage
+
+    def attack(self, other):
+        if isinstance(other, SpaceShip):
+            damage_dealt = self.damage - other.armour
+            if damage_dealt < 0 < self.damage:
+                damage_dealt = 1
+            other.hp = other.hp - damage_dealt
+        print(
+            f'Damage dealt to {other.name} = {self.damage} ({self.name} Damage) - {other.armour} ({other.name} Armor) = {damage_dealt}')
+        print(f'HP left on {other.name} = {other.hp}')
+
+    def upgrade(self, upgrade_name):
+        if upgrade_name == 'Titanium Armour':
+            self.armour = self.armour + 3
+            self.hp = self.hp + 250
+        elif upgrade_name == 'Absorption Shield':
+            self.armour = math.floor(self.armour * 1.5)
+        elif upgrade_name == 'Proton Torpedos':
+            self.damage = self.damage + 25
+        elif upgrade_name == 'Flare Engine':
+            self.damage = math.floor(self.damage * 1.5)
+            self.armour = self.armour + 3
+            self.hp = math.floor(self.hp * 1.2)
+
+
+def main():
+    ship1 = SpaceShip('Nostromo', random.randint(1, 100), random.randint(0, 30), random.randint(0, 30))
+    ship2 = SpaceShip('Covenant', random.randint(1, 100), random.randint(0, 30), random.randint(0, 30))
+
+    print(f'Stats before upgrades:\n'
+          f'Ship 1: Name = {ship1.name}, HP = {ship1.hp}, Armour = {ship1.armour}, Damage = {ship1.damage}\n'
+          f'Ship 2: Name = {ship2.name}, HP = {ship2.hp}, Armour = {ship2.armour}, Damage = {ship2.damage}')
+
+    ship1_upgrades = list()
+    ship2_upgrades = list()
+    while True:
```

Comment:
This is a very ugly way of writing a loop. It will work but it's not readable and it's hard to follow! This is an obvious code smell for me.

When you deal with situations like these, always find a way to write the code to have true end conditions obvious if possible and try to avoid break statements unless they greatly simplify your code.